### ISSUE TITLE
Empty grid stuck in loading state

### ISSUE THREAD

**Issue by paodb** [vaadin/flow-components#4776]
### Description

An empty grid is added to a view, and after all view loading is finished, it can be observed that the the grid has the loading attribute set.

This seems similar to an old issue https://github.com/vaadin/vaadin-grid-flow/issues/1082

### Expected outcome

An empty grid should not have the loading attribute.

### Minimal reproducible example

```
 public EmptyGridView() {
        setSizeFull();

        var grid = new Grid<>(Person.class, false);
        grid.setMultiSort(true);
        grid.setColumnReorderingAllowed(true);
        grid.setSizeFull();
        grid.setItems();

        Grid.Column<Person> firstNameColumn = grid.addColumn(Person::getFirstName)
            .setKey(UUID.randomUUID().toString())
            .setHeader("First name")
            .setSortable(true);

        ArrayList<GridSortOrder<Person>> sortOrder = new ArrayList<>();
        sortOrder.add(new GridSortOrder<>(firstNameColumn, SortDirection.ASCENDING));
        grid.sort(sortOrder);

        add(grid);
    }

}

class Person {
    private int id;
	private String firstName;

    public int getId() {
		return id;
	}

	public void setId(int id) {
		this.id = id;
	}

	public String getFirstName() {
		return firstName;
	}

	public void setFirstName(String firstName) {
		this.firstName = firstName;
	}
```

### Steps to reproduce

1. Add the snipped above to a view.
2. Run the application and observe the empty grid to be rendered.
3. Check the DOM in the browser console and see loading attribute is set 

![image](https://user-images.githubusercontent.com/30666649/223452699-5ed4cec9-5147-469f-8c8c-8b1fd894edbd.png)

### Environment

Vaadin version(s): 23.3.6/24.0.0.rc2

### Browsers

Issue is not browser related

**Comment by yuriy-fix** [comment]
The issue is added to the backlog priority queue for further investigation.

**Comment by bwajtr** [comment]
I've investigated the issue, and I discovered the following:
- The problem is that when you have such an empty grid, then calling `grid.clearCache()` would set the `loading` attribute, but this attribute never goes away, because the `rootPageCallbacks[page]` callback is never executed -> in a very similar fashion as in https://github.com/vaadin/vaadin-grid-flow/issues/1082. 
- the issue is reproducible with the provided minimal example only on 23.3 and 24.0, but not on 24.1
- It's not reproducible on main thanks to [this performance optimization PR](https://github.com/vaadin/flow-components/pull/4774) by @tomivirkki. That optimization removed a call to `grid.clearCache()` when the grid is rendered for the first time -> which fixes (or hides?) _by accident_ also this bug. 
- So one possible way to go is to backport the performance optimization to 23.3 and 24.0 -> but that would really just hide the issue
- Issue is reproducible on 24.1 by calling `grid.clearCache()` manually on client -> I'm not sure how probable this is, but it's a public method so it's at least possible
- I've prepared an integration test that reproduces the issue. Can be found in the following PR: https://github.com/vaadin/flow-components/pull/5038
- However, I'm afraid that a proper fix for this issue is way above my level... there are multiple ways how to do it, and I don't know the guts of gridConnector at a level where I can decide which path to take. I proposed in the same draft PR as above what I think _could be_ a way how to solve this (essentially - if the empty array is returned by the data provider, then put such an empty array to the connector cache too - following mantra that "empty value is a value too, so it's worth caching"). However, when I do it, other problems appear (like: have an empty grid and set items to it via setItems(item1, item2) -> items do not render).
- I'm returning this bug to the pool... will discuss with DS team who could fix it
